### PR TITLE
refactor: [ANDROSDK-1754] ConfigCase from Java/RxJava to Kotlin/Coroutines

### PR DESCRIPTION
**Pull request by dan-dhis2**
refactor:
- ConfigCase.java to ConfigCase.kt
- Remove RxJava from ConfigCase main calls

### CODE REVIEW

File: core/src/main/java/org/hisp/dhis/android/core/sms/domain/interactor/ConfigCase.kt [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+     * Method to download metadata ids. This is required before using the SMS module.
+     */
+    suspend fun refreshMetadataIds() {
+
+        val enabled = localDbRepository.isModuleEnabled.await()
+
+        if (enabled) {
+            try {
+                val config = metadataDownloadConfig.await()
+                val metadata = webApiRepository.getMetadataIds(config).await()
```
Comment (vgarciabnz) [inline core/src/main/java/org/hisp/dhis/android/core/sms/domain/interactor/ConfigCase.kt:149]: This is the API call, which is still a Single. We should get down to API level. The idea behind this refactor is to remove RxJava dependencies at API level.

File: core/src/main/java/org/hisp/dhis/android/core/sms/domain/interactor/ConfigCase.kt [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+        /**
+         * Get the metadata download configuration.
+         * @return `Single with the` metadata download configuration.
+         */
+        get() = localDbRepository.metadataDownloadConfig
+
+    /**
+     * Method to download metadata ids. This is required before using the SMS module.
+     */
+    suspend fun refreshMetadataIds() {
```
Comment (vgarciabnz) [inline core/src/main/java/org/hisp/dhis/android/core/sms/domain/interactor/ConfigCase.kt:142]: This is part of the public API, which should not be modified. It should return a `Completable`.

File: core/src/main/java/org/hisp/dhis/android/core/sms/domain/interactor/ConfigCase.kt [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+            Log.d(TAG, "Not refreshing SMS metadata, because sms module is disabled")
+        }
+
+    }
+
+    /**
+     * Callable that triggers the method [.refreshMetadataIds].
+     * @return Callable object to refresh metadata ids.
+     */
+    suspend fun refreshMetadataIdsCallable() {
```
Comment (vgarciabnz) [inline core/src/main/java/org/hisp/dhis/android/core/sms/domain/interactor/ConfigCase.kt:165]: This is part of the internal API, we can use `suspend` here and modify the method with `internal`.

File: core/src/main/java/org/hisp/dhis/android/core/sms/domain/interactor/ConfigCase.kt [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+     * @param metadataIdsConfig Configuration with the metadata ids.
+     * @return `Completable` that completes when the configuration is changed.
+     */
+    fun setMetadataDownloadConfig(metadataIdsConfig: GetMetadataIdsConfig?): Completable {
+        return if (metadataIdsConfig == null) {
+            Completable.error(IllegalArgumentException("Received null config"))
+        } else localDbRepository.setMetadataDownloadConfig(metadataIdsConfig)
+    }
+
+    val metadataDownloadConfig: Single<GetMetadataIdsConfig>
```
Comment (vgarciabnz) [inline core/src/main/java/org/hisp/dhis/android/core/sms/domain/interactor/ConfigCase.kt:132]: By default, functions starting with `fun getX()` are translated to `val x`, which follows Java conventions but is a bit weird... What do you think? Should we stick to `fun getMetadataDownloadConfig()`?

File: core/src/main/java/org/hisp/dhis/android/core/sms/domain/interactor/ConfigCase.kt [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+import org.hisp.dhis.android.core.sms.domain.repository.internal.LocalDbRepository
+
+/**
+ * Used to set initial data that is common for all sms sending tasks
+ */
+class ConfigCase(
+    private val webApiRepository: WebApiRepository,
+    private val localDbRepository: LocalDbRepository
+) {
+    val smsModuleConfig: Single<SmsConfig>
```
Comment (vgarciabnz) [inline core/src/main/java/org/hisp/dhis/android/core/sms/domain/interactor/ConfigCase.kt:45]: Same comment as below for `getMetadataDownloadConfig`

**Review by vgarciabnz — CHANGES_REQUESTED** [review]
ConfigCase is part of the public API, so their methods' signature should remain the same. The only internal method is `refreshMetadataIdsCallable()`, which is used internally.

File: core/src/main/java/org/hisp/dhis/android/core/sms/domain/interactor/ConfigCase.kt (hunk 4 above)
Comment (dan-dhis2) [reply core/src/main/java/org/hisp/dhis/android/core/sms/domain/interactor/ConfigCase.kt:132]: Agree, the `fun` approach seems more solid and clear

**Comment by dan-dhis2** [comment]
This PR has been reverted to **Draft**, as it still needs analysis and some problems arose with `rxCompletable`

**Review by vgarciabnz — APPROVED** [review]

**Pull request merged**